README.md
# deno-snowflake

![Tests](https://github.com/n1c00o/deno-snowflake/workflows/testing/badge.svg?branch=main)
[![nest badge](https://nest.land/badge.svg)](https://nest.land/package/deno_snowflake)
[![deno doc](https://doc.deno.land/badge.svg)](https://doc.deno.land/https/deno.land/x/deno_snowflake/snowflake.ts)

❄ Deno module to work with Twitter's Snowflake technology

## Explication

Snowflake is, according to the [official and archived repository](https://github.com/twitter-archivesnowflake/tree/snowflake),
is a service for generating unique ID numbers at high scale.

### Original Twitter's Snowflake format

[Snowflake repository](https://github.com/twitter-archive/snowflake/tree/snowflake-2010#solution).

Here is a representation of the default format I took from bwmarrin's [repository](https://github.com/bwmarrin/snowflake)

```js
+--------------------------------------------------------------------------+
| 1 Bit Unused | 41 Bit Timestamp |  10 Bit NodeID  |   12 Bit Sequence ID |
+--------------------------------------------------------------------------+
```

A Snowflake ID following the default settings:

- The ID is a **63 bit integer** stored in an int 64.
- The first **41 bits** are used to store the timestamp _(in milliseconds)_,
  using a [custom epoch](#custom-epoch) _(This provide 69 years of lifetime)_.
- The next **10 bits** contains a Node ID _(a range from 0 to 1023, so 2^10 nodes)_.
- The last **12 bits** corresponds to a sequence number,
  which rolls over every 4096. This sequence is generally used
  as the increment number, a number incremented for every generated ID on the node.

The default settings allow the generation of 4096 unique IDs every millisecond,
per Node ID.

### Custom epoch

The Snowflake's epoch can me modified by the entity using the Snowflake technology,
which means you can use your own.

Here a list of some Snowflake's epoch:

- Twitter Epoch _(used by default by the module)_:
  **1288834974657** -> Nov 04 2010 01:42:54
- Discord Epoch: **1420070400000** -> Jan 01 2015 00:00:00

### Alternative

[sony/sonyflake](https://github.com/sony/sonyflake) is an alternative of
Twitter's Snowflakes which provide more lifetime _(174 years instead of 69)_,
can work on more nodes _(up to 2^16 nodes instead of 2^10)_ but can generate
less ID numbers per millisecond than Snowflakes _(102.4 instead of 2^12 per node)_.

## Getting Started

### Requirements

You need [deno](https://deno.land/) to use this deno module.

### Usage

Import the module to use it.

```ts
import * as Snowflake from 'https://x.nest.land/deno_snowflake@1.0.1/snowflake.ts';
// import * as Snowflake from "https://deno.land/x/deno_snowflake/snowflake@v1.0.1.ts";
```

## Documentation

### melt

```ts
export function melt(snowflakeID: string, epoch: number): MeltedSnowflake;
```

This function deconstructs a Snowflake ID and returns its information,
such as the timestamp.

If no `epoch` provided, the module use the exported constant [`TWITTER_EPOCH`](#TWITTER_EPOCH).

### generate

```ts
export function generate(options: SnowflakeIDGeneratorOptions): string;
```

This function generate a new Snowflake ID using given options and returns it.

If no `options` provided, the module use the following properties:

```ts
{
  timestamp: Date.now(),
  epoch: TWITTER_EPOCH,
  processID: 0,
  workerID: 1,
}
```

### TWITTER_EPOCH

```ts
export const TWITTER_EPOCH = 1288834974657;
```

The Epoch Twitter uses for Snowflakes generation.
This value is used by default in the [`melt`](#melt) function.

### sequence

```ts
export let sequence = 0;
```

The sequence used in the [`generate`](#generate) function.
The variable should not be modified and its exported only for reading the value.

### MeltedSnowflake

```ts
export interface MeltedSnowflake {
	timestamp: number;
	processID: number;
	workerID: number;
	sequence: number;
	epoch: number;
	binary: string;
	stringID: string;
	bigIntID: bigint;
	date: Date;
}
```

Type of the object returned by the [`melt`](#melt) function.

- `timestamp`: Timestamp the Snowflake was created.
- `processID`: The ID of the process that generate the Snowflake ID.
- `workerID`: The ID of the worker that generate the Snowflake ID.
- `sequence`: Sequence in the Snowflake.
- `epoch`: The used Epoch in the Snowflake.
- `binary`: Binary representation of the Snowflake ID.
- `stringID`: The Snowflake ID as a string.
- `bigIntID`: The Snowflake ID as a BigInt.
- `date`: The Date the Snowflake was created.

## License

The module is licensed under the MIT License.
Please read [LICENSE](LICENSE) for more information.

## Contributing

Thanks for wanting to contribute to `deno_snowflake`!
Please read [CONTRIBUTING.md](CONTRIBUTING.md) and [CODE_OF_CONDUCT](CODE_OF_CONDUCT.md).


CHANGELOG.md
# deno-snowflake's CHANGELOG

## 1.0.0 - HEAD


snowflake_test.ts
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";
import { generate, melt, MeltedSnowflake, TWITTER_EPOCH } from "./snowflake.ts";

Deno.test({
  // Information comes from https://developer.twitter.com/en/docs/twitter-api/v1/tweets/post-and-engage/api-reference/get-statuses-show-id
  // The tweet with ID 1050118621198921728 has a timestamp that should be 1539202764211

  name:
    "Get information about Snowflake ID `1050118621198921728` with Twitter Epoch (`1288834974657`)",
  fn(): void {
    const melted: MeltedSnowflake = melt("1050118621198921728", TWITTER_EPOCH);
    assertEquals(String(melted.timestamp), "1539202764211");
  },
});

Deno.test({
  name: "Generate a Snowflake ID",
  fn(): void {
    const timestamp: number = Date.now();
    const snowflakeID = generate({
      timestamp,
      epoch: TWITTER_EPOCH,
      processID: 0,
      workerID: 1,
    });
    const meltedSnowflakeID: MeltedSnowflake = melt(snowflakeID, TWITTER_EPOCH);
    assertEquals(String(meltedSnowflakeID.timestamp), String(timestamp));
  },
});


snowflake.ts
/**
 * Twitter Epoch, used as the default epoch. (1288834974657 -> Nov 04 2010 01:42:54)
 */
export const TWITTER_EPOCH = 1288834974657;

/** The sequence to generate a Snowflake ID. Become 0 when the value reach 4096.*/
export let sequence = 0;

/** Represents a deconstructed Snowflake */
export interface MeltedSnowflake {
  /** Timestamp the Snowflake was created */
  timestamp: number;
  /** The ID of the process that generate the Snowflake ID. */
  processID: number;
  /** The ID of the worker that generate the Snowflake ID. */
  workerID: number;
  /** Sequence in the Snowflake */
  sequence: number;
  /** The used Epoch in the Snowflake */
  epoch: number;
  /** Binary representation of the Snowflake ID */
  binary: string;
  /** The Snowflake ID as a string */
  stringID: string;
  /** The Snowflake ID as a BigInt */
  bigIntID: bigint;
  /** The Date the Snowflake was created */
  date: Date;
}

/** Options to generate a Snowflake ID */
export interface SnowflakeIDGeneratorOptions {
  /** The timestamp of the Snowflake ID generation. By default, we recommend `Date.now` */
  timestamp?: number;
  /** The Epoch used for the Snowflake ID generation. By default, we use the Twitter Epoch (`1288834974657`) */
  epoch?: number;
  /** The ID of the process that generate the Snowflake ID. By default this is 0 */
  processID?: number;
  /** The ID of the worker that generate the Snowflake ID. By default this is 1 */
  workerID?: number;
}

/** Deconstructs (melts) a Snowflake ID and returns its information */
export function melt(
  snowflakeID: string,
  epoch: number = TWITTER_EPOCH,
): MeltedSnowflake {
  const binarySnowflakeID = BigInt(snowflakeID).toString(2).padStart(64, "0");

  const snowflakeTimestamp: number =
    parseInt(binarySnowflakeID.substring(0, 42), 2) + epoch;

  return {
    stringID: snowflakeID,
    bigIntID: BigInt(snowflakeID),
    epoch,
    timestamp: snowflakeTimestamp,
    workerID: parseInt(binarySnowflakeID.substring(42, 47), 2),
    processID: parseInt(binarySnowflakeID.substring(47, 52), 2),
    sequence: parseInt(binarySnowflakeID.substring(52, 64), 2),
    binary: binarySnowflakeID,
    date: new Date(snowflakeTimestamp),
  };
}

/** Generate a Snowflake ID and returns it */
export function generate(options: SnowflakeIDGeneratorOptions = {
  timestamp: Date.now(),
  epoch: TWITTER_EPOCH,
  processID: 0,
  workerID: 1,
}): string {
  // The sequence should never bypass 4095 since the variable varies
  // from 0 to 4095 (to allow the creation of 2^12 different ID numbers per ms)
  sequence++;
  if (sequence > 4095) sequence = 0;
  const snowflakeID = String(
    (((BigInt(options.timestamp || Date.now()) -
      BigInt(options.epoch || TWITTER_EPOCH)) <<
      22n) +
      BigInt(options.processID || 0) +
      BigInt(options.workerID || 1)) +
      BigInt(sequence),
  );

  return snowflakeID;
}
